Initial commit of CMake build system

dnakamura opened the pull request:
Adds experimental support for building with cmake

Issue: #933
Signed-off-by: Robert Young <rwy0717@gmail.com>
Signed-off-by: Devin Nakamura <devinn@ca.ibm.com>

rwy0717 commented:
If you take a look at the CI, you'll see some changes to the way things are being built. The big one is an experimental build with CMake on x86-64, which is allowed to fail. We've also moved to the "Trusty" Ubuntu distribution. Since the OMRChecker isn't building on Trusty (yet), there is a new build on Precise for running the linter.

As well, there is a new Appveyor configuration using CMake.

mgaudet commented:
Would you be able to send me a link to one of the logs that failed with the linter on Trusty so I can get a fix in?

Leonardo2718 commented on fvtest/CMakeLists.txt:
```diff
@@ -0,0 +1,13 @@
+add_subdirectory(omrGtestGlue)
+add_subdirectory(algotest)
+# add_subdirectory(compilertest)
+add_subdirectory(gctest)
```
Would it be possible to also add `jitbuildertest` to the list?

rwy0717 replied:
Yep! We can't build the test yet, but we can leave a note.

rwy0717 commented:
@mgaudet [here](https://travis-ci.org/rwy0717/omr/jobs/220064747) is a link to a failed Trusty linter build.

The clang and llvm versions are output in the `build-on-travis.sh` script:
```
+llvm-config --version
3.5.0
+clang++ --version
clang version 3.5.0 (tags/RELEASE_350/final)
Target: x86_64-unknown-linux-gnu
Thread model: posix
+make lint
```

fjeremic commented on omrversionstrings.CMakeTemplate.h:
```diff
@@ -0,0 +1,29 @@
+/*******************************************************************************
+ *
+ * (c) Copyright IBM Corp. 2015
```
This needs a copyright update. Is this an tool-generated file?

rwy0717 replied:
This is a written file that is used to generate `omrversionstrings.h`. Will update the copyright.

rwy0717 commented:
This is good to go.

fjeremic commented:
General question: Are the two build systems going to be running concurrently for a while?

dnakamura commented:
@fjeremic  Yes the two build systems will run side by side for the time being.  This pr adds an allowed failure cmake build to the travis config.

fjeremic commented:
Anything blocking this to be merged? Are we waiting for additional reviews from @youngar and @mstoodle?

dnakamura commented:
@fjeremic  It's my understanding that  we are waiting on @mstoodle  and the eclipse CQ process

mstoodle commented:
Thanks for the extra @ mention, I had missed the original request to review. Will review by EOD.

mstoodle commented:
First observation: the commit message is pretty terse :) .

mstoodle commented on .appveyor.yml:
```diff
@@ -0,0 +1,36 @@
+###############################################################################
+#
+# (c) Copyright IBM Corp. 2016
```
new file, we're 4 months into 2017 already :)

mstoodle commented on .gitignore:
```diff
@@ -16,6 +16,15 @@
 #    Multiple authors (IBM Corp.) - initial implementation and documentation
 ###############################################################################
```
general comment: modified files should have their "rightmost" copyright date updated to the current year

mstoodle commented on CMakeLists.txt:
```diff
@@ -0,0 +1,115 @@
+###############################################################################
+#
+# (c) Copyright IBM Corp. 2016
```
should be 2017

mstoodle commented on CMakeLists.txt:
```diff
@@ -0,0 +1,115 @@
+
+cmake_minimum_required(VERSION 3.2 FATAL_ERROR)
+
+set(OMR_VERSION_MAJOR 0)
```
would it be better for this version identification piece to live in a separate file, isolated from the rest of the cmake machinery?

dnakamura replied:
It would appear this is how LLVM does it. I find they are usually a good standard for accepted practice.
https://github.com/llvm-mirror/llvm/blob/master/CMakeLists.txt

dnakamura replied:
Also, now that I think about it, this is really more of a place holder. Ideally we need figure out a system where we can store version information once and consume in both build systems

mstoodle commented on cmake/config.cmake:
```diff
@@ -0,0 +1,102 @@
+
```
is this a generated file? is that why there is no copyright header?

mstoodle commented on cmake/modules/CMakeASM-ZOSInformation.cmake:
```diff
@@ -0,0 +1,9 @@
+set(CMAKE_ASM-ZOS_SOURCE_FILE_EXTENSIONS s)
```
no copyright header?

mstoodle commented on cmake/modules/CMakeDetermineASM-ZOSCompiler.cmake:
```diff
@@ -0,0 +1,4 @@
+set(ASM_DIALECT "-ZOS")
```
no copyright header?

mstoodle commented on cmake/modules/CMakeTestASM-ZOSCompiler.cmake:
```diff
@@ -0,0 +1,3 @@
+set(ASM_DIALECT "-ZOS")
```
no copyright header?

mstoodle commented on cmake/platform.cmake:
```diff
@@ -0,0 +1,264 @@
+
```
no copyright header?

mstoodle commented on cmake/platform.cmake:
```diff
@@ -0,0 +1,264 @@
+	set(CMAKE_RUNTIME_OUTPUT_DIRECTORY "${CMAKE_BINARY_DIR}")
+endif()
+
+if(OMR_HOST_OS STREQUAL "zos")
```
(chose arbitrary "if()" to put this comment)

Longer term, we probably want to organize this platform-specific code better than lumped together in a single piece of if(some platform) hell.

Of course, there's platform-specific if()'s all over the place anyway, so I guess it's really a general cleanup issue to be implemented at some point.

mstoodle commented on example/CMakeLists.txt:
```diff
@@ -0,0 +1,30 @@
+###############################################################################
+#
+# (c) Copyright IBM Corp. 2016
```
should be 2017

mstoodle commented on example/glue/CMakeLists.txt:
```diff
@@ -0,0 +1,36 @@
+###############################################################################
+#
+# (c) Copyright IBM Corp. 2016
```
should be 2017

mstoodle commented on fvtest/CMakeLists.txt:
```diff
@@ -0,0 +1,32 @@
+###############################################################################
+#
+# (c) Copyright IBM Corp. 2016
```
2017

mstoodle commented on fvtest/algotest/CMakeLists.txt:
```diff
@@ -0,0 +1,75 @@
+###############################################################################
+#
+# (c) Copyright IBM Corp. 2016
```
2017

mstoodle commented on fvtest/gctest/CMakeLists.txt:
```diff
@@ -0,0 +1,61 @@
+###############################################################################
+#
+# (c) Copyright IBM Corp. 2016
```
2017

mstoodle commented on fvtest/omrGtestGlue/CMakeLists.txt:
```diff
@@ -0,0 +1,48 @@
+###############################################################################
+#
+# (c) Copyright IBM Corp. 2016
```
2017

mstoodle commented on fvtest/porttest/CMakeLists.txt:
```diff
@@ -0,0 +1,107 @@
+###############################################################################
+#
+# (c) Copyright IBM Corp. 2016
```
2017

mstoodle commented on fvtest/sigtest/CMakeLists.txt:
```diff
@@ -0,0 +1,77 @@
+###############################################################################
+#
+# (c) Copyright IBM Corp. 2016
```
2017

mstoodle commented on fvtest/sigtest/CMakeLists.txt:
```diff
@@ -0,0 +1,77 @@
+)
+
+
+#MODULE_INCLUDES += ../util
```
broad review comment: please document reason for having all these (and others, where you haven't documented TODO's) blocks of statements commented out.

I tried to identify it where ever I saw it in an attempt to help you, but probably still missed a few spots.

mstoodle commented on fvtest/threadtest/CMakeLists.txt:
```diff
@@ -0,0 +1,103 @@
+###############################################################################
+#
+# (c) Copyright IBM Corp. 2016
```
2017

mstoodle commented on fvtest/threadtest/CMakeLists.txt:
```diff
@@ -0,0 +1,103 @@
+	threadTestHelp.cpp
+)
+
+#ifeq (1,$(OMR_THR_FORK_SUPPORT))
```
please document why these statements are commented out

mstoodle commented on fvtest/util/CMakeLists.txt:
```diff
@@ -0,0 +1,24 @@
+###############################################################################
+#
+# (c) Copyright IBM Corp. 2016
```
2017

mstoodle commented on fvtest/utiltest/CMakeLists.txt:
```diff
@@ -0,0 +1,39 @@
+###############################################################################
+#
+# (c) Copyright IBM Corp. 2016
```
2017

mstoodle commented on gc/CMakeLists.txt:
```diff
@@ -0,0 +1,279 @@
+###############################################################################
+#
+# (c) Copyright IBM Corp. 2016
```
2017

mstoodle commented on glue/CMakeLists.txt:
```diff
@@ -0,0 +1,48 @@
+###############################################################################
+#
+# (c) Copyright IBM Corp. 2016
```
2017

mstoodle commented on omr/CMakeLists.txt:
```diff
@@ -0,0 +1,50 @@
+###############################################################################
+#
+# (c) Copyright IBM Corp. 2016
```
2017

mstoodle commented on omrcfg.CMakeTemplate.h:
```diff
@@ -0,0 +1,225 @@
+/*******************************************************************************
+ *
+ * (c) Copyright IBM Corp. 2017
```
yay!! :)

mstoodle commented on omrsigcompat/CMakeLists.txt:
```diff
@@ -0,0 +1,95 @@
+###############################################################################
+#
+# (c) Copyright IBM Corp. 2016
```
2017

mstoodle commented on omrsigcompat/CMakeLists.txt:
```diff
@@ -0,0 +1,95 @@
+	omrsig.cpp
+)
+
+#TODO need more comrehensive symbols
```
tupo: comprehensive

mstoodle commented on omrsigcompat/CMakeLists.txt:
```diff
@@ -0,0 +1,95 @@
+#EXPORT_FUNCTIONS_FILE := omrsig.exportlist
+
+### Create platform specific exports
+
```
please document why commented out

mstoodle commented on omrtrace/CMakeLists.txt:
```diff
@@ -0,0 +1,36 @@
+###############################################################################
+#
+# (c) Copyright IBM Corp. 2016
```
2017

mstoodle commented on omrtrace/CMakeLists.txt:
```diff
@@ -0,0 +1,36 @@
+)
+
+#ifeq (gcc,$(OMR_TOOLCHAIN))
+## disable builtin strncpy buffer length check for optimized builds (_FORTIFY_SOURCE=1 requires -O3)
```
please document why commented out

mstoodle commented on port/CMakeLists.txt:
```diff
@@ -0,0 +1,358 @@
+###############################################################################
+#
+# (c) Copyright IBM Corp. 2016
```
2017

mstoodle commented on scripts/build-on-travis.sh:
```diff
@@ -0,0 +1,29 @@
+#!/bin/bash
```
copyright header?

mstoodle commented on third_party/CMakeLists.txt:
```diff
@@ -0,0 +1,20 @@
+###############################################################################
+#
+# (c) Copyright IBM Corp. 2016
```
2017

mstoodle commented on third_party/pugixml-1.5/CMakeLists.txt:
```diff
@@ -0,0 +1,26 @@
+###############################################################################
+#
+# (c) Copyright IBM Corp. 2016
```
2017

mstoodle commented on thread/CMakeLists.txt:
```diff
@@ -0,0 +1,296 @@
+###############################################################################
+#
+# (c) Copyright IBM Corp. 2016
```
2017

mstoodle commented on thread/CMakeLists.txt:
```diff
@@ -0,0 +1,296 @@
+	rwmutex.c
+)
+
+#ifeq (1,$(OMR_THR_JLM))
```
please document why commented out

mstoodle commented on thread/CMakeLists.txt:
```diff
@@ -0,0 +1,296 @@
+	include_directories(osx unix)
+endif()
+
+#ifeq (aix,$(OMR_HOST_OS))
```
please document why commented out

mstoodle commented on thread/CMakeLists.txt:
```diff
@@ -0,0 +1,296 @@
+#endif
+
+list(APPEND VPATHS common)
+#MODULE_INCLUDES += $(THREAD_SRCDIR)common
```
more commented out statements; please document why

mstoodle commented on tools/CMakeLists.txt:
```diff
@@ -0,0 +1,76 @@
+###############################################################################
+#
+# (c) Copyright IBM Corp. 2016
```
2017

mstoodle commented on tools/hookgen/CMakeLists.txt:
```diff
@@ -0,0 +1,28 @@
+###############################################################################
+#
+# (c) Copyright IBM Corp. 2016
```
2017

mstoodle commented on tools/tracegen/CMakeLists.txt:
```diff
@@ -0,0 +1,50 @@
+###############################################################################
+#
+# (c) Copyright IBM Corp. 2016
```
2017

mstoodle commented on tools/tracemerge/CMakeLists.txt:
```diff
@@ -0,0 +1,32 @@
+###############################################################################
+#
+# (c) Copyright IBM Corp. 2016
```
2017

mstoodle commented on util/CMakeLists.txt:
```diff
@@ -0,0 +1,24 @@
+###############################################################################
+#
+# (c) Copyright IBM Corp. 2016
```
2017

mstoodle commented on util/avl/CMakeLists.txt:
```diff
@@ -0,0 +1,30 @@
+###############################################################################
+#
+# (c) Copyright IBM Corp. 2016
```
2017

mstoodle commented on util/hashtable/CMakeLists.txt:
```diff
@@ -0,0 +1,43 @@
+###############################################################################
+#
+# (c) Copyright IBM Corp. 2016
```
2017

mstoodle commented on util/hashtable/CMakeLists.txt:
```diff
@@ -0,0 +1,43 @@
+###############################################################################
+
+
+#TODO:
```
please document why commented out

mstoodle commented on util/hookable/CMakeLists.txt:
```diff
@@ -0,0 +1,27 @@
+###############################################################################
+#
+# (c) Copyright IBM Corp. 2016
```
2017

mstoodle commented on util/hookable/CMakeLists.txt:
```diff
@@ -0,0 +1,27 @@
+#    Multiple authors (IBM Corp.) - initial implementation and documentation
+###############################################################################
+
+#TODO there is a bunch of stuff with vpaths, presumably for some extensibility reasons
```
this TODO comment doesn't really describe an action, could you please make it more clear what needs to be done here?

mstoodle commented on util/omrutil/CMakeLists.txt:
```diff
@@ -0,0 +1,105 @@
+###############################################################################
+#
+# (c) Copyright IBM Corp. 2016
```
2017

mstoodle commented on util/omrutil/CMakeLists.txt:
```diff
@@ -0,0 +1,105 @@
+#    Multiple authors (IBM Corp.) - initial implementation and documentation
+###############################################################################
+
+#TODO
```
please document why statements are commented out and what the TODO is

mstoodle commented on util/pool/CMakeLists.txt:
```diff
@@ -0,0 +1,30 @@
+###############################################################################
+#
+# (c) Copyright IBM Corp. 2016
```
2017

mstoodle commented on util/pool/CMakeLists.txt:
```diff
@@ -0,0 +1,30 @@
+	ut_pool.c
+)
+
+#target_include_directories(j9pool
```
please document why statements are commented out

mstoodle commented on CMakeLists.txt:
```diff
@@ -0,0 +1,115 @@
+    set(output "")
+  foreach(currentFile ${${inlist}})
+    set(fileFound False)
+    #message("Searching for ${currentFile}")
```
if this diagnostic output is useful, maybe it would be better to introduce an option to control the output rather than just commenting out message lines.

mstoodle commented on cmake/platform.cmake:
```diff
@@ -0,0 +1,264 @@
+message(STATUS "CMAKE_SYSTEM_PROCESSOR=${CMAKE_SYSTEM_PROCESSOR}")
+message(STATUS "CMAKE_CROSSCOMPILING=${CMAKE_CROSSCOMPILING}")
+
+#set(OMR_ARCH_POWER Off Internal "Power CPU Architecture")
```
are these commented out statements the fundamental spot to enable platforms that aren't yet supported by our cmake build implementation? If so, please leave a comment to that effect. If it's not, please leave a comment explaining what they are for.

dnakamura replied:
I believe these are superfluous. Actual set moved to inside the if() statements

mstoodle commented on cmake/platform.cmake:
```diff
@@ -0,0 +1,264 @@
+
+	set(CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} ${linker_common} /INCREMENTAL:NO /NOLOGO /LARGEADDRESSAWARE wsetargv.obj")
+	if(OMR_ENV_DATA64)
+		#TODO: makefile has this but it seems to break shit
```
please remove the profanity

mstoodle commented on cmake/platform.cmake:
```diff
@@ -0,0 +1,264 @@
+	else()
+		set(CMAKE_SHARED_LINKER_FLAGS "${CMAKE_SHARED_LINKER_FLAGS} -entry:_DllMainCRTStartup@12")
+	endif()
+	#set(CMAKE_C_FLAGS "${common_flags}")
```
please document why commented out

mstoodle commented on cmake/platform.cmake:
```diff
@@ -0,0 +1,264 @@
+
+	# # This is the first option applied to the C++ linking command.
+	# # It is not applied to the C linking command.
+	# OMR_MK_CXXLINKFLAGS=-Wc,"langlvl(extended)" -+
```
please document why commented out

mstoodle commented on fvtest/gctest/CMakeLists.txt:
```diff
@@ -0,0 +1,61 @@
+	omrutil
+	omrglue
+)
+#	./omrgctest -configListFile=fvtest/gctest/configuration/fvConfigListFile.txt
```
please document why these lines are commented out

mstoodle commented on fvtest/omrGtestGlue/CMakeLists.txt:
```diff
@@ -0,0 +1,48 @@
+
+target_link_libraries(omrGtestGlue INTERFACE omrGtest)
+
+#TODO there is platform specific stuff here
```
doesn't sound like an action to me, could you please make this comment more clear?

mstoodle commented:
Lest it be lost in the massive dump of review comments, I think this PR represents some amazing work, and I'm *really* happy to see it. Thanks @dnakamura and @rwy0717 for putting it all together!

mstoodle commented:
Just curious, what motivated the switch to use Ubuntu Trusty?

dnakamura commented:
@mstoodle  thanks for the feedback

- Most of the commented out blocks are chunks of the source makefile which have not yet been ported over (will update to reflect this)

- If I recall correctly the main reason to move to trusty was that precise has some horridly out of date version of cmake (2.8.7)

dnakamura commented:
@mstoodle  I've addressed the comments you have made. I've left it as a separate commit in order to make it easier to track changes, but will squash it done before we merge in

0xdaryl commented:
@marti4d : FYI, you may want to take a peek at this

mstoodle commented on thread/CMakeLists.txt:
```diff
@@ -42,6 +42,7 @@ list(APPEND OBJECTS
 	rwmutex.c
 )
 
+#TODO need to add te
```
te? either a typo or please explain a bit more

dnakamura commented:
@mstoodle  should be good to go

youngar commented:
I opened an Eclipse CQ for this pull request [here](https://dev.eclipse.org/ipzilla/show_bug.cgi?id=13286). I'll update here with any CQ status updates.

rwy0717 commented:
I've rebased against latest, and added a new tracegen definition file, `j9hook.tdf`, to the build. The cmake build should be passing again.

youngar commented:
The Eclipse CQ has received "PMC approval", but we are still waiting for a preliminary legal review before we can check in this code.  I'm guessing the review will be completed sometime tomorrow.  You can read more about this process [here](https://wiki.eclipse.org/Development_Resources/HOWTO/Parallel_IP_Process).

charliegracie commented:
The CQ has not been approved for release yet.  I will keep monitoring it

charliegracie commented:
CQ has been completely approved :).

charliegracie commented:
Discussed with @youngar that if it got approved while he was away I would handle merging.
